pkg/util/configmap_util.go

```diff
@@ -145,92 +145,67 @@ func UpdateConfigMapWithKeyValue(ctx context.Context, cm *corev1.ConfigMap, key,
 	if err != nil {
-		return nil, err
+		return err
 	}
 	klog.Infof("NodeStageVolume successfully stored lock info {%s: %s} in configmap %s/%s with new data %v", key, value, updatedCM.Namespace, updatedCM.Name, updatedCM.Data)
-	return updatedCM, nil
-}
-
-// UpdateConfigMapWithData sets configmap.data to the given map, and updates the configmap in the api server.
-// No-op if no changes to configmap.data.
-// Returns the server's representation of the configMap, and an error, if there is any.
-func UpdateConfigMapWithData(ctx context.Context, cm *corev1.ConfigMap, data map[string]string, client kubernetes.Interface) (*corev1.ConfigMap, error) {
-	// No-op if no changes to configmap.
-	if reflect.DeepEqual(cm.Data, data) && containsFinalizer(cm, ConfigMapFinalzer) {
-		return cm, nil
-	}
-	cm.Data = data
-	addFinalizer(cm, ConfigMapFinalzer)
-	updatedCM, err := client.CoreV1().ConfigMaps(cm.Namespace).Update(ctx, cm, metav1.UpdateOptions{})
-	if err != nil {
-		return nil, err
-	}
-	return updatedCM, nil
-}
-
-// DeleteConfigMap removes the finalizer, and deletes the configmap from the api server.
-func DeleteConfigMap(ctx context.Context, cm *corev1.ConfigMap, client kubernetes.Interface) error {
-	if containsFinalizer(cm, ConfigMapFinalzer) {
-		removeFinalizer(cm, ConfigMapFinalzer)
-		if _, err := client.CoreV1().ConfigMaps(cm.Namespace).Update(ctx, cm, metav1.UpdateOptions{}); err != nil {
-			if apiError.IsNotFound(err) {
-				return nil
-			}
-			return err
-		}
-	}
-	if cm.DeletionTimestamp == nil {
-		if err := client.CoreV1().ConfigMaps(cm.Namespace).Delete(ctx, cm.Name, metav1.DeleteOptions{}); err != nil {
-			if apiError.IsNotFound(err) {
-				return nil
-			}
-			return err
-		}
-	}
 	return nil
 }
 
-// RemoveKeyFromConfigMap deletes the key from configmap.data.
+// RemoveKeyFromConfigMap deletes the key from configmap.data, then updates the configmap.
 // No-op if the key does not exist.
-// Delete the configmap from api server if configmap.data is an empty map.
-// Otherwise, update the configmap.
-func RemoveKeyFromConfigMap(ctx context.Context, cm *corev1.ConfigMap, key string, client kubernetes.Interface) (*corev1.ConfigMap, error) {
+func RemoveKeyFromConfigMap(ctx context.Context, cm *corev1.ConfigMap, key string, client kubernetes.Interface) error {
 	_, cmNeedsUpdate := cm.Data[key]
 	if !cmNeedsUpdate {
 		klog.Infof("NodeUnstageVolume skipped updating configmap %s/%s since key %s not found in configmap.data", cm.Namespace, cm.Name, key)
-		return cm, nil
+		return nil
 	}
 
 	klog.Infof("NodeUnstageVolume removing key %s from configmap %s/%s with data %v", key, cm.Namespace, cm.Name, cm.Data)
 	delete(cm.Data, key)
-	if len(cm.Data) == 0 {
-		klog.Infof("NodeUnstageVolume deleting configmap %s/%s since configmap.data is empty", cm.Namespace, cm.Name)
-		return nil, DeleteConfigMap(ctx, cm, client)
-	}
 	addFinalizer(cm, ConfigMapFinalzer)
```

Comment:
why are we adding finalizer on updates? should it be always there during the first creation of the configMap?

Reply:
Resolving offline, finalizer will only be added during creation